jgrapht-core/src/main/java/org/jgrapht/alg/tour/NearestInsertionHeuristicTSP.java

```diff
@@ -0,0 +1,330 @@
+ * </p>
+ *
+ * <p>
+ * This implementation can also be used in order to try to improve an existing
+ * tour. See method {@link #improveTour(GraphPath)}.
+ * </p>
+ *
+ * <p>
+ * The runtime complexity of this class is $O(V^2)$.
+ * </p>
+ *
+ * <p>
+ * This algorithm requires that the graph is complete.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Peter Harman
+ */
+public class NearestInsertionHeuristicTSP<V, E> extends
+        HamiltonianCycleAlgorithmBase<V, E>
+        implements HamiltonianCycleImprovementAlgorithm<V, E> {
+
+    private GraphPath<V, E> subtour;
+
+    /**
+     * Constructor. By default a sub-tour is chosen based on the shortest edge
+     */
+    public NearestInsertionHeuristicTSP() {
+        this(null);
+    }
+
+    /**
+     * Constructor
+     *
+     * @param subtour Initial sub-tour, or null to start with shortest edge
+     */
+    public NearestInsertionHeuristicTSP(GraphPath<V, E> subtour) {
+        this.subtour = subtour;
+    }
+
+    /**
+     * Computes a tour using the nearest insertion heuristic.
+     *
+     * @param graph the input graph
+     * @return a tour
+     * @throws IllegalArgumentException if the graph is not undirected
+     * @throws IllegalArgumentException if the graph is not complete
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        // Check that graph is appropriate
+        checkGraph(graph);
+
+        if (graph.vertexSet().size() == 1) {
+            return getSingletonTour(graph);
+        }
+        return vertexListToTour(augment(subtour(graph), graph), graph);
+    }
+
+    /**
+     * Improves an existing tour. Using this algorithm will not improve the
+     * cost, but instead adds any missing vertices to the tour. It also removes
+     * vertices from the tour if not in the graph, so this can be used to update
+     * a tour following changes to a graph.
+     *
+     * @param tour The existing tour
+     * @return A valid and complete tour for the same graph
+     * @throws IllegalArgumentException if the graph is not undirected
+     * @throws IllegalArgumentException if the graph is not complete
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> improveTour(GraphPath<V, E> tour) {
+        // Check that graph is appropriate
+        checkGraph(this.subtour.getGraph());
+
+        this.subtour = tour;
+        Graph<V, E> graph = this.subtour.getGraph();
+        return vertexListToTour(augment(subtour(graph), graph), graph);
+    }
+
+    /**
+     * Get or create a sub-tour to start augmenting
+     *
+     * @param graph The graph
+     * @return Vertices of an initial sub-tour
+     */
+    private List<V> subtour(Graph<V, E> graph) {
+        List<V> subtourVertices = new ArrayList<>();
+        if (subtour != null) {
```

Comment:
I think this filtering logic should be removed because the algorithm shouldn't care whether the graph is modified or not. Graph modification is application-specific. 

Currently, if the path of length one is supplied, it is ignored because of this filtering. I think the important logic is: 
* if we get an initial path, it is our start point
* if we don't get any path, we choose the shortest edge

We should not take responsibility for the graph modifications because we might guess wrongly how the graph is modified (existing algorithms have the same logic).

Reply:
Done